presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -0,0 +1,72 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.JoinNode.DistributionType;
+
+import static com.facebook.presto.SystemSessionProperties.isDistributedJoinEnabled;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isAtMostScalar;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+
+public class DetermineJoinDistributionType
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(joinNode -> !joinNode.getDistributionType().isPresent());
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        DistributionType distributionType = determineDistributionType(node, context.getSession(), context.getLookup());
+        return Result.ofPlanNode(node.withDistributionType(distributionType));
+    }
+
+    public static DistributionType determineDistributionType(JoinNode node, Session session, Lookup lookup)
+    {
+        // The implementation of full outer join only works if the data is hash partitioned.
+        JoinNode.Type type = node.getType();
+        if (type == RIGHT || type == FULL || (isDistributedJoinEnabled(session) && !mustBroadcastJoin(node, lookup))) {
+            return PARTITIONED;
+        }
+
+        return REPLICATED;
+    }
+
+    private static boolean mustBroadcastJoin(JoinNode node, Lookup lookup)
+    {
+        return isAtMostScalar(node.getRight(), lookup) || isCrossJoin(node);
+    }
+
+    private static boolean isCrossJoin(JoinNode node)
+    {
+        return node.getType() == INNER && node.getCriteria().isEmpty();
```

Comment:
use `JoinNode#isCrossJoin` instead.

Reply:
You might want to update `com.facebook.presto.sql.planner.optimizations.DetermineJoinDistributionType.Rewriter#isCrossJoin`

Reply:
`JoinNode#isCrossJoin` returns `true` when join doesn't have equi criteria but just a filter. It would instruct the rule to pick the partitioned join which makes no sense.